parlai/core/teachers.py

```diff
@@ -1944,6 +1945,223 @@ def shutdown(self):
+            self.chunks = shared['chunks']
+            self.samples = shared['samples']
+            self.rng = shared['rng']
+            self.is_root_teacher = False
+        else:
+            self.is_root_teacher = True
+            self.samples = queue.Queue(maxsize=self.buffersize)
+            self.chunks = queue.Queue()
+            if self.is_train:
+                self.rng = random.Random()
+            else:
+                self.rng = random.Random(4)
+            self._enqueue_chunks()
+            # launch queue loader on the main thread
+            self._enqueue_request()
+
+        self.episode_done = True
+
+    def _get_data_folder(self):
+        if not self.opt.get('datafile'):
+            raise RuntimeError(
+                'Must specify datafile or override this function '
+                'to return the data folder.'
+            )
+
+        return self.opt['datafile']
+
+    @abstractmethod
+    def _get_num_samples(self, datatype) -> int:
+        """
+        [Abstract] Return the number of samples given the datatype.
+        """
+        pass
+
+    @abstractmethod
+    def _get_fold_chunks(self, datatype) -> List[int]:   # type: ignore
+        """
+        [Abstract] Return a list of chunk IDs (integer).
+
+        Given the datatype (train/test/valid), return the list of
+        chunk IDs that correspond to that split.
+        """
+        pass
+
+    def set_datasettings(self, datatype):
+        self.folder = self._get_data_folder()
+        self.num_samples = self._get_num_samples(datatype)
+        self.fold_chunks = self._get_fold_chunks(datatype)
+
+        self.is_train = 'train' in datatype and 'evalmode' not in datatype
+
+    def share(self):
+        shared = super().share()
+        shared['samples'] = self.samples
+        shared['chunks'] = self.chunks
+        shared['rng'] = self.rng
+        return shared
+
+    def _setup_data(self, datatype):
+        pass
+
+    def num_episodes(self):
+        return self.num_samples
+
+    def num_examples(self):
+        return self.num_samples
+
+    def _enqueue_request(self):
+        """
+        Queue a request for loading to the data loader.
+        """
+        self.data_loader.request_load(self.receive_data, self._get_chunk, ())
+
+    def receive_data(self, future):
+        """
+        Loads data.
+
+        Load data into self.samples until buffersize is
+        reached.
+        """
+        data = future.result()
+        if data is None:
+            logging.debug('No longer queueing')
+            return
+        while data:
+            # self.samples is a queue with maxsize
+            # self.buffersize, so will block if the
+            # buffer gets full
+            self.samples.put(data.pop(0))
+        # and start loading the next chunk
+        self._enqueue_request()
+
+    def _enqueue_chunks(self):
+        """
+        Shuffles and queues fold chunks for loading.
+        """
+        self.rng.shuffle(self.fold_chunks)
```

Comment:
I think we have to protect against this if it's valid/test/train:evalmode. They should produce the same outputs every time.